scripts/solc_project.py

```diff
@@ -0,0 +1,102 @@
+#!/usr/bin/env python3
+
+"""
+Authors: Tobias Kaiser, Jakob Beckmann
+
+Copyright 2018 ChainSecurity AG
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+import os
+import sys
+from pathlib import Path
+import json
+import subprocess
+
+from solc import install_solc
+from solc.main import _parse_compiler_output
+from solc.wrapper import solc_wrapper
+from solc.exceptions import SolcError
+import solc.install
+
+from . import project
+from . import utils
+
+class SolcProject(project.Project):
+    """A project that uses the `solc` compiler."""
+    def __init__(self, project_root):
+        super().__init__(project_root)
+
+    def compile_(self):
+        """Compile the project and dump the output to an intermediate file."""
+        sources = self._get_sol_files()
+        if not sources:
+            raise utils.NoSolidityProject(self.get_project_root())
+        comp_output = self._compile_solfiles(sources)
+        with open(self.get_compilation_output(), 'w') as fs:
+            json.dump(comp_output, fs)
+
+    def _get_sol_files(self):
+        """Returns the solidity files contained in the project root."""
+        return [os.path.join(p, f) for p, _, fs in os.walk(self.get_project_root()) for f in fs if
+                f.endswith('.sol') and
+                'node_modules' not in p and
+                '/test/' not in p[len(str(self.get_project_root())):] and
+                not p.endswith('/test')]
+
+    def _get_binary(self, version):
+        """Returns the binary for some version of solc."""
+        binary = os.path.join(Path.home(), f'.py-solc/solc-v{version}/bin/solc')
+        if not os.path.exists(binary):
+            # install in seperate process to control stdout
+            cmd = ["python3", "-c", f"from solc import install_solc\ninstall_solc('v{version}', platform='linux')"]
+            process = subprocess.run(cmd, shell=False, stdout=subprocess.PIPE)
```

Comment:
I'm not sure I understand the need for this subprocess, running this python code directly does not work?

Reply:
It does. The reason I used subprocess is to suppress the output from the install unless the install fails. The reason is that optimally Securify should only output the violations/warnings/compliances, but if an install occurs, running the code directly would print _a lot_ of clutter to the console. The only way I found to suppress this was using subprocess.